### PR TITLE
pycrumbs 0.0.1

### PR DESCRIPTION
**Pull request by mpacula**
Long overdue first release based on a rewrite of our biobank data challenge submission. See examples/example01.py for a short demo script.

[Build Status](https://travis-ci.org/MGHComputationalPathology/pycrumbs)

### CODE REVIEW

File: pycrumbs/layout.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    :return: list of dimensions
+    """
+    def shift_all(dimensions, dx, dy):
+        """Shifts all dimnensions by the same amount"""
+        return [Dimensions(dim.node, dim.x + dx, dim.y + dy, dim.width, dim.height,
+                           node_x=dim.node_x + dx, node_y=dim.node_y + dy)
+                for dim in dimensions]
+
+    def bounding_box(dimensions):
+        """Computers the total dimensions (bounding box) of all dimensions in a list"""
```
Comment (alliemclean) [inline pycrumbs/layout.py]: Super minor.  "Computers" is probable a typo.

File: pycrumbs/__init__.py [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+    :param spring_constant: spring constant for force layout.
+    :param parent_multiplier: strength multiplier for force from parent
+    :param neutral_length: neutral spring length. Default 1.0
+    :return: None
+
+    """
+
+    def default_get_node_size(node):
+        """Default node size: area-preserving wrt number of entities"""
+        return size * math.sqrt(1.0 * len(node.entities) / len(tree.entities))  # area is linearly proportional
```
Comment (alliemclean) [inline pycrumbs/__init__.py:322]: size and tree are not local variables.  Is this intended?

File: pycrumbs/__init__.py [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+    :param parent: parent node
+    :param child: child node
+    :param max_observations: for nodes with multiple associated trajectories, the new "observation" will be a union
+    e.g. "OBS1 or OBS2 or OBS3". max_observations specifies the max number of alternatives before the result
+    is abbreviated as "other". Default 1
+
+    """
+    def begins_with(lst, prefix):
+        """True iff lst begins with a prefix"""
+        if len(lst) < len(prefix):
```
Comment (alliemclean) [inline pycrumbs/__init__.py:274]: I am assuming this is for strings.  Why not use startswith()?

Nevermind... I just noticed it's not for strings... maybe add a comment that it's meant for trajectories.

**Review by alliemclean — APPROVED** [review]

File: pycrumbs/__init__.py (hunk 2 above)
Comment (mpacula) [reply pycrumbs/__init__.py:322]: Yes, they are parameters to the parent function which never change, so easier to create a closure than to pass them around.

File: pycrumbs/__init__.py (hunk 3 above)
Comment (mpacula) [reply pycrumbs/__init__.py:274]: Good point, I'll add a comment to clarify.

File: pycrumbs/layout.py (hunk 1 above)
Comment (mpacula) [reply pycrumbs/layout.py]: Grate catch, thanks

**Pull request merged**